livepeer/studio
# api: Create authorization API to be used by other services

**What does this pull request do? Explain your changes. (required)**
This is to create a new `/api/auth` endpoint which can be used as an API for authorizing
requests in other services in our infrastructure. The design was strongly based on 
[NGINX'es sub-request authorization](https://docs.nginx.com/nginx/admin-guide/security-controls/configuring-subrequest-authentication), so that the authorization request is as much as a "proxy"
as possible, adding only additional headers to account for additional information that may
be useful for the authorization process.

In other words, the authorization request has:
 - The same method as the original request
 - No body (authorization should be based on the accessed resource only, not the contents of the request)
 - All the original headers from the request (but we only really need the `Authorization` one, so services may optimize)
 - An `X-Original-Uri` header with... the original URI of the request being authorized
 - An `X-Livepeer-Stream-Id` header with the ID of the stream which is being accessed. This is used to 
 detect cross-account accesses which should not be authorized.

And the repsonse of the `/api/auth` request can also be directly proxied to the client, as it will contain
 whatever verbatim authorization error should be given to the client.

**Specific updates (required)**
 - Add support in the `auth` middleware for an `originalUriHeader`, which should be used instead of
 the URI for authorizing a given request.
    - The middleware is responsible mostly for authentication, and authorization with the built-in 
    API key auth access rules (`X-Original-Uri` header).
 - Create a new `auth` controller and register its API on our router.
    - The controller is responsible for the authorization of access to the given resource. This currently only
    means that the stream ID belongs to the calling user (or the user is an admin).

## -

- **How did you test each of these updates (required)**
 - Start API and analyzer configured to authorize with the API
 - Try making calls without an auth header and check it fails
   - Same with invalid auth headers
 - Check that a user can access their own stream
 - Check that a user cannot access other users streams
 - Check that an admin can access any stream
All good ✅ 

**Does this pull request close any open issues?**
Implements the first part of https://github.com/livepeer/livepeer-data/issues/4

**Checklist:**
